### PR TITLE
Gp/build speepud

### PR DESCRIPTION
**Pull request by JackPiri**
To be answered:

- how to make sure `node/build.rs` runs when appropriate (always?)?
- is it fine to use global `config.toml` instead than local `config.toml`?

**EDIT:**

**Decisions taken:**

- global `config.toml` is used (inside `CARGO_HOME`)
- `node/build.rs` runs whenever `config.toml` is modified

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -18,3 +18,7 @@ coverage_summary.txt
 
 # Ignore output reports
 *_output.txt
+
+# Ignore external libraries (and their paths)
+deps/
+#config.toml
```
[the remark this answers is not in the document]
Comment (JackPiri) [reply .gitignore:24]: Yes, it was kept before deciding to go for `CARGO_HOME/config.toml`. Now removed.

File: node/build.rs [hunk 2]
```diff
@@ -13,10 +13,119 @@
...
+    libs[0].1 && libs[1].1
+}
+
+fn set_env_paths(reset: bool) {
+    let libs_path: PathBuf = env::current_dir().unwrap().join("../deps");
+    let libs_path = libs_path.to_str().unwrap();
+    let cargo_config = PathBuf::from(env!("CARGO_HOME")).join("config.toml");
+    // let cargo_config = PathBuf::from(env::current_dir().unwrap())
+    //     .join("..")
+    //     .join("config.toml");
```
[the remark this answers is not in the document]
Comment (JackPiri) [reply node/build.rs:91]: Yes, it was kept before deciding to go for CARGO_HOME/config.toml. Now removed.

File: node/build.rs [hunk 3]
```diff
@@ -13,10 +13,119 @@
...
+
+    if !Path::new(&cargo_config).exists() {
+        let _ = File::create(cargo_config.clone());
```
[the remark this answers is not in the document]
Comment (JackPiri) [reply node/build.rs:94]: Ok replaced all `let _ ` occurrences with `.expect()`.

File: node/build.rs [hunk 4]
```diff
@@ -13,10 +13,119 @@
...
+    }
+
+    let mut file = File::open(cargo_config.clone()).unwrap();
```
[the remark this answers is not in the document]
Comment (JackPiri) [reply node/build.rs:97]: Ok modified.

File: node/build.rs [hunk 5]
```diff
@@ -13,10 +13,119 @@
...
+        for target_path in target_paths {
+            for entry in WalkDir::new(target_path).into_iter().flatten() {
+                let path = entry.path();
+                for lib in libs.iter_mut() {
+                    let regex_pattern = Regex::new(&lib.0).unwrap();
+                    if path.is_file() && regex_pattern.is_match(path.to_str().unwrap()) {
+                        let mut destination_file = String::from(destination_path);
+                        destination_file.push_str(path.file_name().unwrap().to_str().unwrap());
+                        let _ = fs::create_dir_all(destination_path);
+                        let _ = fs::copy(path, destination_file);
```
[the remark this answers is not in the document]
Comment (JackPiri) [reply node/build.rs:68]: Ok replaced all let _  occurrences with .expect().

File: node/build.rs [hunk 6]
```diff
@@ -13,10 +13,119 @@
...
+                        lib.1 = true;
+                        break;
+                    }
```
[the remark this answers is not in the document]
Comment (JackPiri) [reply node/build.rs:71]: Ok modified.

File: node/build.rs [hunk 7]
```diff
@@ -13,10 +13,119 @@
...
+            format!("{}{}", "librocksdb-sys-.*/out/", rocksdb_lib),
+            false,
+        ),
+        (format!("{}{}", "librocksdb-sys-.*/out/", snappy_lib), false),
+    ];
+    let destination_path = "../deps/";
+
+    if fs::metadata(PathBuf::from(destination_path).join(rocksdb_lib)).is_err()
+        || fs::metadata(PathBuf::from(destination_path).join(snappy_lib)).is_err()
+    {
```
[the remark this answers is not in the document]
Comment (JackPiri) [reply node/build.rs:58]: From the docs of `fs::matadata` I read that one of the errors condition is `path does not exist`; here I want to check if the libs are already cached/copied, if so just skip all the rest of the function.

File: node/Cargo.toml [hunk 8]
```diff
@@ -74,6 +74,9 @@ try-runtime-cli = { workspace = true, optional = true }
 
 [build-dependencies]
 substrate-build-script-utils = { workspace = true }
+toml = "0.8.14"
```
[the remark this answers is not in the document]
Comment (JackPiri) [reply node/Cargo.toml:77]: Mhm unfortunately this is not a one-line change, the struct and interfaces used by the two crates are different. If it's not a problem I would leave `toml` crate.

**Review by la10736 — APPROVED** [review]
GREAT

**Review by DanieleDiBenedetto — APPROVED** [review]

**Pull request merged**